packages/@aws-cdk/cx-api/lib/context/vpc.ts

```diff
@@ -1,4 +1,4 @@
-export const VPC_PROVIDER = 'vpc-provider';
+export const VPC_PROVIDER = 'vpc-provider2';
```

Comment:
This seems dangerous. It means the CLI cannot possibly recognize both `vpc-provider1` and `vpc-provider2` requests at the same time, which means the CLI breaks backwards compatibility with old CX-api versions.

How do we make sure that the new CLI will still handle old frameworks?

Reply:
Damn, you're right. That means we have to have a new provider, just for this case.

Will change.

Reply:
Here's an interesting question.

We will have 2 very similar providers: the "old" VPC one, and the "new" VPC one that accepts asymmetric subnets. Their code will be quite similar. Should I:

a) have the new provider be completely separate from the old one - they should not share any code (as the old one is there only for backwards compatibility reasons); or

b) should I refactor the old one to have an abstract class, and then have both the old one and the new one inherit from this abstract class, to reduce code duplication?

@rix0rrr what do you think?

Reply:
Maybe even have a single class and have them do something else based on the input format? Instead of adding a new provider you could also add a version number in the request.

But yes, otherwise let's share code. Maybe the V1 uses the V2, and applies additional restrictions on the output afterwards?

Also, we're going to need to bump the CX version number to tell people that they need to upgrade their CLI to use the new VPC provider (don't forget to add a "version upgrade" code path, which again can do nothing except push up the CX version number one).

Reply:
> Maybe even have a single class and have them do something else based on the input format? Instead of adding a new provider you could also add a version number in the request.

I like this idea. That's what I did in the newest version. There's a new optional property for the request, `returnAsymmetricSubnets`, which `Vpc.fromLookup()` now uses, and it triggers the new behavior in the (existing) VPC provider.

> Also, we're going to need to bump the CX version number to tell people that they need to upgrade their CLI to use the new VPC provider (don't forget to add a "version upgrade" code path, which again can do nothing except push up the CX version number one).

Can you expand what needs to be done here? I don't think I understood your comment fully. Is it just changing [this](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/cx-api/lib/versioning.ts#L34-L34) to `1.15.0`, or is there something else to be done here?

Reply:
https://github.com/aws/aws-cdk/blob/c0c0513b68cc12e6696c57d2305bf84311ad9c59/packages/%40aws-cdk/cx-api/lib/versioning.ts#L19

Reply:
Done here: https://github.com/aws/aws-cdk/pull/4544/files?file-filters%5B%5D=.ts#diff-eacf6cc8a5af37f2f5e8202fca7d8ec3R67-R70 , I hope I did it correctly.